### PR TITLE
Improve the gRPC BBEs

### PR DESCRIPTION
**Pull request by BuddhiWathsala**
## Purpose
$subject

## Test Environment 
Tested with the latest generated pack(SLP9-SNAPSHOT)

### CODE REVIEW

File: examples/grpc-client-streaming/grpc_client_streaming_service.bal [hunk 1]
```diff
@@ -1,33 +1,54 @@
...
 }
-service /HelloWorld on new grpc:Listener(9090) {
-
-    isolated function lotsOfGreetings(grpc:Caller caller,
-                            stream<string,error> clientStream) {
-        log:print("Client connected successfully.");
-        //Read and process each message in the client stream
+service "HelloWorld" on ep {
+    remote function lotsOfGreetings(stream<string,error> clientStream) returns string|error {
+        io:println("connected sucessfully.");
```
Comment (ThisaruGuruge) [inline examples/grpc-client-streaming/grpc_client_streaming_service.bal:13]: Aren't we supposed to use `log` in services and `io` in clients?

File: examples/grpc-bidirectional-streaming/grpc_bidirectional_streaming_service.bal [hunk 2]
```diff
@@ -1,54 +1,67 @@
...
-                } else {
-                    log:print("Server message to caller " + callerId
-                                                    + " sent successfully.");
-                }
-            }
+service "Chat" on ep {
+    remote function chat(stream<ChatMessage, error> clientStream) returns stream<string> {
+        string[] responses = [];
+        int i = 0;
+        // Read and process each message in the client stream
```
Comment (praneesha) [inline examples/grpc-bidirectional-streaming/grpc_bidirectional_streaming_service.bal:15]:
```suggestion
        // Read and process each message in the client stream.
```

File: examples/grpc-bidirectional-streaming/grpc_bidirectional_streaming_service.bal [hunk 3]
```diff
@@ -1,54 +1,67 @@
...
-                } else {
-                    log:print("Server message to caller " + callerId
-                                                    + " sent successfully.");
-                }
-            }
-        //If the client sends an error to the server, the stream closes and returns the error
-        } else if (e is error) {
-            log:printError("Error from Connector: " + e.message());
-        }
+        // Once the client sends a notification to indicate the end of the stream, 'grpc:EOS' is returned by the stream
```
Comment (praneesha) [inline examples/grpc-bidirectional-streaming/grpc_bidirectional_streaming_service.bal:22]:
```suggestion
        // Once the client sends a notification to indicate the end of the stream, 'grpc:EOS' is returned by the stream.
```

File: examples/grpc-bidirectional-streaming/grpc_bidirectional_streaming_service.bal [hunk 4]
```diff
@@ -1,54 +1,67 @@
...
+        return self.caller->sendError(response);
+    }
+
+    isolated remote function complete() returns grpc:Error? {
+        return self.caller->complete();
     }
 }
+
+
+// Context record includes message payload and headers.
```
Comment (praneesha) [inline examples/grpc-bidirectional-streaming/grpc_bidirectional_streaming_service.bal:52]:
```suggestion
// The `context` record includes the message payload and headers.
```

File: examples/grpc-bidirectional-streaming/grpc_bidirectional_streaming_service.bal [hunk 5]
```diff
@@ -1,54 +1,67 @@
...
+public type ContextString record {|
+    stream<string> content;
+    map<string[]> headers;
+|};
+
+// Context record includes message payload and headers.
```
Comment (praneesha) [inline examples/grpc-bidirectional-streaming/grpc_bidirectional_streaming_service.bal:58]:
```suggestion
// The `context` record includes the message payload and headers.
```

File: examples/grpc-bidirectional-streaming/grpc_bidirectional_streaming_service_client.bal [hunk 6]
```diff
@@ -1,57 +1,77 @@
 // This is client implementation for bidirectional streaming scenario.
```
Comment (praneesha) [inline examples/grpc-bidirectional-streaming/grpc_bidirectional_streaming_service_client.bal:1]:
```suggestion
// This is the client implementation of the bidirectional streaming scenario.
```

File: examples/grpc-bidirectional-streaming/grpc_bidirectional_streaming_service.bal [hunk 7]
```diff
@@ -1,54 +1,67 @@
 // This is the server implementation for the bidirectional streaming scenario.
```
Comment (praneesha) [inline examples/grpc-bidirectional-streaming/grpc_bidirectional_streaming_service.bal:1]:
```suggestion
// This is the server implementation of the bidirectional streaming scenario.
```

File: examples/grpc-bidirectional-streaming/grpc_bidirectional_streaming_service_client.bal [hunk 8]
```diff
@@ -1,57 +1,77 @@
...
 import ballerina/grpc;
 import ballerina/io;
-import ballerina/lang.runtime;
 
-int total = 0;
-public function main() {
+public function main (string... args) returns error? {
+    // Client endpoint configuration
```
Comment (praneesha) [inline examples/grpc-bidirectional-streaming/grpc_bidirectional_streaming_service_client.bal:6]:
```suggestion
    // Client endpoint configuration.
```

File: examples/grpc-bidirectional-streaming/grpc_bidirectional_streaming_service_client.bal [hunk 9]
```diff
@@ -1,57 +1,77 @@
...
+    ChatClient ep = check new("http://localhost:9090");
+    // Executes the RPC call and receives the customized streaming client
```
Comment (praneesha) [inline examples/grpc-bidirectional-streaming/grpc_bidirectional_streaming_service_client.bal:8]:
```suggestion
    // Executes the RPC call and receives the customized streaming client.
```

File: examples/grpc-bidirectional-streaming/grpc_bidirectional_streaming_service_client.bal [hunk 10]
```diff
@@ -1,57 +1,77 @@
...
+    // Sends multiple messages to the server.
+    ChatMessage[] messages = [
+        {name: "Sam", message: "Hi"},
+        {name: "Ann", message: "Hey"},
+        {name: "John", message: "Hello"}
+    ];
+    foreach ChatMessage msg in messages {
+        check streamingClient->send(msg);
+    }
+    // Once all messages are sent, client send complete message to notify the server, I’m done.
```
Comment (praneesha) [inline examples/grpc-bidirectional-streaming/grpc_bidirectional_streaming_service_client.bal:20]:
```suggestion
    // Once all the messages are sent, the client sends the message to notify the server about the completion.
```

File: examples/grpc-bidirectional-streaming/grpc_bidirectional_streaming_service_client.bal [hunk 11]
```diff
@@ -1,57 +1,77 @@
...
+    check streamingClient->complete();
+    // Receives the server stream response iteratively
```
Comment (praneesha) [inline examples/grpc-bidirectional-streaming/grpc_bidirectional_streaming_service_client.bal:22]:
```suggestion
    // Receives the server stream response iteratively.
```

File: examples/grpc-bidirectional-streaming/grpc_bidirectional_streaming_service_client.bal [hunk 12]
```diff
@@ -1,57 +1,77 @@
...
 
-    // Sends multiple messages to the server.
-    ChatMessage mes = {name: "Sam", message: "Hi "};
-    grpc:Error? connErr = ep->send(mes);
+    private grpc:Client grpcClient;
 
-    if (connErr is grpc:Error) {
-        io:println("Error from Connector: " + connErr.message());
+    public isolated function init(string url, grpc:ClientConfiguration? config = ()) returns grpc:Error? {
+        // Initialize client endpoint.
```
Comment (praneesha) [inline examples/grpc-bidirectional-streaming/grpc_bidirectional_streaming_service_client.bal:37]:
```suggestion
        // Initialize the client endpoint.
```

File: examples/grpc-client-streaming/grpc_client_streaming_service.bal [hunk 13]
```diff
@@ -1,33 +1,54 @@
 // This is the server implementation for the client streaming scenario.
```
Comment (praneesha) [inline examples/grpc-client-streaming/grpc_client_streaming_service.bal:1]:
```suggestion
// This is the server implementation of the client streaming scenario.
```

File: examples/grpc-client-streaming/grpc_client_streaming_service.bal [hunk 14]
```diff
@@ -1,33 +1,54 @@
...
-service /HelloWorld on new grpc:Listener(9090) {
-
-    isolated function lotsOfGreetings(grpc:Caller caller,
-                            stream<string,error> clientStream) {
-        log:print("Client connected successfully.");
-        //Read and process each message in the client stream
+service "HelloWorld" on ep {
+    remote function lotsOfGreetings(stream<string,error> clientStream) returns string|error {
+        io:println("connected sucessfully.");
+        // Read and process each message in the client stream
```
Comment (praneesha) [inline examples/grpc-client-streaming/grpc_client_streaming_service.bal:14]:
```suggestion
        // Read and process each message in the client stream.
```

File: examples/grpc-client-streaming/grpc_client_streaming_service.bal [hunk 15]
```diff
@@ -1,33 +1,54 @@
...
         error? e = clientStream.forEach(isolated function(string name) {
-            log:print("Server received greet: " + name);
+            io:println("greet received: ", name);
         });
-
         //Once the client sends a notification to indicate the end of the stream, 'grpc:EOS' is returned by the stream
```
Comment (praneesha) [inline examples/grpc-client-streaming/grpc_client_streaming_service.bal:18]:
```suggestion
        //Once the client sends a notification to indicate the end of the stream, 'grpc:EOS' is returned by the stream.
```

File: examples/grpc-client-streaming/grpc_client_streaming_service.bal [hunk 16]
```diff
@@ -1,33 +1,54 @@
...
+    isolated remote function sendError(grpc:Error response) returns grpc:Error? {
+        return self.caller->sendError(response);
+    }
+
+    isolated remote function complete() returns grpc:Error? {
+        return self.caller->complete();
     }
 }
+
+// Context record includes message payload and headers
```
Comment (praneesha) [inline examples/grpc-client-streaming/grpc_client_streaming_service.bal:50]:
```suggestion
// The `context` record includes the message payload and headers.
```

File: examples/grpc-client-streaming/grpc_client_streaming_service_client.bal [hunk 17]
```diff
@@ -2,62 +2,67 @@
 import ballerina/grpc;
 import ballerina/io;
 
-int total = 0;
-public function main() {
+public function main (string... args) returns error? {
     // Client endpoint configuration.
```
Comment (praneesha) [inline examples/grpc-client-streaming/grpc_client_streaming_service_client.bal:6]:
```suggestion
    // The client endpoint configuration.
```

File: examples/grpc-client-streaming/grpc_client_streaming_service_client.bal [hunk 18]
```diff
@@ -2,62 +2,67 @@
...
-    HelloWorldClient helloWorldEp = new ("http://localhost:9090");
+    HelloWorldClient ep = check new("http://localhost:9090");
+    string[] requests = ["Hi Sam", "Hey Sam", "GM Sam"];
+    // Execute the client-streaming RPC call and receives the streaming client
```
Comment (praneesha) [inline examples/grpc-client-streaming/grpc_client_streaming_service_client.bal:9]:
```suggestion
    // Execute the client-streaming RPC call and receive the streaming client.
```

File: examples/grpc-client-streaming/grpc_client_streaming_service_client.bal [hunk 19]
```diff
@@ -2,62 +2,67 @@
...
+    LotsOfGreetingsStreamingClient streamingClient = check ep->lotsOfGreetings();
+    // Send multiple messages to the server
```
Comment (praneesha) [inline examples/grpc-client-streaming/grpc_client_streaming_service_client.bal:11]:
```suggestion
    // Send multiple messages to the server.
```

File: examples/grpc-client-streaming/grpc_client_streaming_service_client.bal [hunk 20]
```diff
@@ -2,62 +2,67 @@
...
+    foreach var greet in requests {
+        checkpanic streamingClient->send(greet);
+    }
+    // Once all the messages are sent, the server notifies the caller with a `complete` message
```
Comment (praneesha) [inline examples/grpc-client-streaming/grpc_client_streaming_service_client.bal:15]:
```suggestion
    // Once all the messages are sent, the server notifies the caller with a `complete` message.
```

File: examples/grpc-client-streaming/grpc_client_streaming_service_client.bal [hunk 21]
```diff
@@ -2,62 +2,67 @@
...
+    io:println("Completed successfully");
+    anydata response = checkpanic streamingClient->receive();
+    io:println(response);
+
+}
 
-    grpc:StreamingClient ep;
-    // Execute the unary non-blocking call that registers a server message listener.
-    var res = helloWorldEp->lotsOfGreetings(HelloWorldMessageListener);
+// The client that used to invoke the RPC
```
Comment (praneesha) [inline examples/grpc-client-streaming/grpc_client_streaming_service_client.bal:23]:
```suggestion
// The client that used to invoke the RPC.
```

File: examples/grpc-client-streaming/grpc_client_streaming_service_client.bal [hunk 22]
```diff
@@ -2,62 +2,67 @@
...
-        return;
-    } else {
-        io:println("Initialized connection sucessfully.");
-        ep = res;
+    *grpc:AbstractClientEndpoint;
+
+    private grpc:Client grpcClient;
+
+    public isolated function init(string url, grpc:ClientConfiguration? config = ()) returns grpc:Error? {
+        // initialize client endpoint.
```
Comment (praneesha) [inline examples/grpc-client-streaming/grpc_client_streaming_service_client.bal:31]:
```suggestion
        // Initialize the client endpoint.
```

File: examples/grpc-client-streaming/grpc_client_streaming_service_client.bal [hunk 23]
```diff
@@ -2,62 +2,67 @@
...
+        grpc:StreamingClient sClient = check self.grpcClient->executeClientStreaming("HelloWorld/lotsOfGreetings");
+        return new LotsOfGreetingsStreamingClient(sClient);
     }
+}
 
-    // Once all the messages are sent, the server notifies the caller with a `complete` message.
-    grpc:Error? result = ep->complete();
-    if (result is grpc:Error) {
-        io:println("Error in sending complete message", result);
+// The streaming client that used send the streaming messages
```
Comment (praneesha) [inline examples/grpc-client-streaming/grpc_client_streaming_service_client.bal:42]:
```suggestion
// The streaming client, which is used to send the streaming messages.
```

File: examples/grpc-secured-unary/grpc_secured_unary_service.bal [hunk 24]
```diff
@@ -1,7 +1,7 @@
 // This is the server implementation for the secured connection (HTTPS) scenario.
```
Comment (praneesha) [inline examples/grpc-secured-unary/grpc_secured_unary_service.bal:1]:
```suggestion
// This is the server implementation of the secured connection (HTTPS) scenario.
```

File: examples/grpc-secured-unary/grpc_secured_unary_service.bal [hunk 25]
```diff
@@ -1,7 +1,7 @@
...
 import ballerina/config;
 import ballerina/grpc;
-import ballerina/log;
+import ballerina/io;
 
 // Server endpoint configuration with the SSL configurations.
```
Comment (praneesha) [inline examples/grpc-secured-unary/grpc_secured_unary_service.bal:6]:
```suggestion
// The server endpoint configuration with the SSL configurations.
```

File: examples/grpc-secured-unary/grpc_secured_unary_service.bal [hunk 26]
```diff
@@ -19,25 +19,45 @@ listener grpc:Listener ep = new (9090, {
...
-
-        if (err is grpc:Error) {
-            log:printError("Error from Connector: " + err.message());
-        } else {
-            log:print("Server send response : " + message);
-        }
+service "HelloWorld" on ep {
+    remote function hello(HelloWorldStringCaller caller, ContextString request) {
+        io:println("Invoked the hello RPC call.");
+        // Reads the request content
```
Comment (praneesha) [inline examples/grpc-secured-unary/grpc_secured_unary_service.bal:25]:
```suggestion
        // Reads the request content.
```

File: examples/grpc-secured-unary/grpc_secured_unary_service.bal [hunk 27]
```diff
@@ -19,25 +19,45 @@ listener grpc:Listener ep = new (9090, {
...
+        string message = "Hello " + request.content;
+        io:println(request);
+        // Set up the response message and send it
```
Comment (praneesha) [inline examples/grpc-secured-unary/grpc_secured_unary_service.bal:28]:
```suggestion
        // Set up the response message and send it.
```

File: examples/grpc-secured-unary/grpc_secured_unary_service.bal [hunk 28]
```diff
@@ -19,25 +19,45 @@ listener grpc:Listener ep = new (9090, {
...
+    isolated remote function sendError(grpc:Error response) returns grpc:Error? {
+        return self.caller->sendError(response);
+    }
+
+    isolated remote function complete() returns grpc:Error? {
+        return self.caller->complete();
     }
 }
+
+# Context record includes message payload and headers.
```
Comment (praneesha) [inline examples/grpc-secured-unary/grpc_secured_unary_service.bal:59]:
```suggestion
# The `context` record includes the message payload and headers.
```

File: examples/grpc-secured-unary/grpc_secured_unary_service_client.bal [hunk 29]
```diff
@@ -3,9 +3,9 @@ import ballerina/config;
 import ballerina/grpc;
 import ballerina/io;
 
-public function main() {
+public function main (string... args) returns error? {
     // Client endpoint configuration with SSL configurations.
```
Comment (praneesha) [inline examples/grpc-secured-unary/grpc_secured_unary_service_client.bal:7]:
```suggestion
    // The client endpoint configuration with the SSL configurations.
```

File: examples/grpc-secured-unary/grpc_secured_unary_service_client.bal [hunk 30]
```diff
@@ -15,14 +15,58 @@ public function main() {
                 }
             }
     });
+    ContextString requestMessage = {content: "WSO2", headers: {}};
+    // Executing unary call
```
Comment (praneesha) [inline examples/grpc-secured-unary/grpc_secured_unary_service_client.bal:19]:
```suggestion
    // Executing the unary call.
```

File: examples/grpc-secured-unary/grpc_secured_unary_service_client.bal [hunk 31]
```diff
@@ -15,14 +15,58 @@ public function main() {
...
+    ContextString result = check ep->helloContext(requestMessage);
+    // Print the content
```
Comment (praneesha) [inline examples/grpc-secured-unary/grpc_secured_unary_service_client.bal:21]:
```suggestion
    // Print the content.
```

File: examples/grpc-secured-unary/grpc_secured_unary_service_client.bal [hunk 32]
```diff
@@ -15,14 +15,58 @@ public function main() {
...
-    if (unionResp is grpc:Error) {
-        io:println("Error from Connector: " + unionResp.message());
-    } else {
-        string result;
-        [result, _] = unionResp;
-        io:println("Client Got Response : " + result);
+    private grpc:Client grpcClient;
+
+    public isolated function init(string url, grpc:ClientConfiguration? config = ()) returns grpc:Error? {
+        // Initialize client endpoint.
```
Comment (praneesha) [inline examples/grpc-secured-unary/grpc_secured_unary_service_client.bal:32]:
```suggestion
        // Initialize the client endpoint.
```

File: examples/grpc-secured-unary/grpc_secured_unary_service_client.bal [hunk 33]
```diff
@@ -15,14 +15,58 @@ public function main() {
...
+            message = req;
+        }
+        var payload = check self.grpcClient->executeSimpleRPC("HelloWorld/hello", message, headers);
+        [anydata, map<string|string[]>][result, respHeaders] = payload;
+        return {content: result.toString(), headers: respHeaders};
+    }
+
 }
+
+# Context record includes message payload and headers.
```
Comment (praneesha) [inline examples/grpc-secured-unary/grpc_secured_unary_service_client.bal:68]:
```suggestion
# The `context` record includes the message payload and headers.
```

File: examples/grpc-server-streaming/grpc_server_streaming_service.bal [hunk 34]
```diff
@@ -1,33 +1,56 @@
 // This is the server implementation for the server streaming scenario.
```
Comment (praneesha) [inline examples/grpc-server-streaming/grpc_server_streaming_service.bal:1]:
```suggestion
// This is the server implementation of the server streaming scenario.
```

File: examples/grpc-server-streaming/grpc_server_streaming_service.bal [hunk 35]
```diff
@@ -1,33 +1,56 @@
...
-
-    isolated function lotsOfReplies(grpc:Caller caller, string name) {
-        log:print("Server received hello from " + name);
+service "HelloWorld" on ep {
+    remote function lotsOfReplies(string name) returns stream<string,error>|error {
+        io:println("Server received hello from " + name);
         string[] greets = ["Hi", "Hey", "GM"];
-
-        // Send multiple messages to the caller.
+        // Create the array of responses by appending recieved name.
```
Comment (praneesha) [inline examples/grpc-server-streaming/grpc_server_streaming_service.bal:15]:
```suggestion
        // Create the array of responses by appending the recieved name.
```

File: examples/grpc-server-streaming/grpc_server_streaming_service.bal [hunk 36]
```diff
@@ -1,33 +1,56 @@
...
+    isolated remote function sendError(grpc:Error response) returns grpc:Error? {
+        return self.caller->sendError(response);
+    }
+
+    isolated remote function complete() returns grpc:Error? {
+        return self.caller->complete();
     }
 }
+
+// Context record includes message payload and headers
```
Comment (praneesha) [inline examples/grpc-server-streaming/grpc_server_streaming_service.bal:50]:
```suggestion
// The `context` record includes the message payload and headers.
```

File: examples/grpc-server-streaming/grpc_server_streaming_service_client.bal [hunk 37]
```diff
@@ -2,40 +2,32 @@
 import ballerina/grpc;
 import ballerina/io;
 
-int total = 0;
-public function main() {
+public function main (string... args) returns error? {
     // Client endpoint configuration.
```
Comment (praneesha) [inline examples/grpc-server-streaming/grpc_server_streaming_service_client.bal:6]:
```suggestion
    // The client endpoint configuration.
```

File: examples/grpc-server-streaming/grpc_server_streaming_service_client.bal [hunk 38]
```diff
@@ -2,40 +2,32 @@
...
-    HelloWorldClient helloWorldEp = new ("http://localhost:9090");
+    HelloWorldClient ep = check new("http://localhost:9090");
+    // Execute the streaming RPC call that registers the server message listener and get the response as a stream.
```
Comment (praneesha) [inline examples/grpc-server-streaming/grpc_server_streaming_service_client.bal:8]:
```suggestion
    // Execute the streaming RPC call that registers the server message listener and gets the response as a stream.
```

File: examples/grpc-server-streaming/grpc_server_streaming_service_client.bal [hunk 39]
```diff
@@ -2,40 +2,32 @@
...
 
-    // Execute the unary non-blocking call that registers the server message listener.
-    grpc:Error? result = helloWorldEp->lotsOfReplies("Sam",
-                                                    HelloWorldMessageListener);
-    if (result is grpc:Error) {
-        io:println("Error from Connector: " + result.message());
-    } else {
-        io:println("Connected successfully");
-    }
+// The client that used to invoke the RPC.
```
Comment (praneesha) [inline examples/grpc-server-streaming/grpc_server_streaming_service_client.bal:16]:
```suggestion
// The client, which is used to invoke the RPC.
```

File: examples/grpc-server-streaming/grpc_server_streaming_service_client.bal [hunk 40]
```diff
@@ -2,40 +2,32 @@
...
 
-// Server Message Listener.
-service object{} HelloWorldMessageListener = service object {
+    private grpc:Client grpcClient;
 
-    // The `resource` registered to receive server messages
-    function onMessage(string message) {
-        io:println("Response received from server: " + message);
+    public isolated function init(string url, grpc:ClientConfiguration? config = ()) returns grpc:Error? {
+        // initialize client endpoint.
```
Comment (praneesha) [inline examples/grpc-server-streaming/grpc_server_streaming_service_client.bal:24]:
```suggestion
        // Initialize the client endpoint.
```

File: examples/grpc-unary-blocking/grpc_unary_blocking_service.bal [hunk 41]
```diff
@@ -1,36 +1,58 @@
 // This is the server implementation for the unary blocking/unblocking scenario.
```
Comment (praneesha) [inline examples/grpc-unary-blocking/grpc_unary_blocking_service.bal:1]:
```suggestion
// This is the server implementation of the unary blocking/unblocking scenario.
```

File: examples/grpc-unary-blocking/grpc_unary_blocking_service.bal [hunk 42]
```diff
@@ -1,36 +1,58 @@
...
 
 @grpc:ServiceDescriptor {
     descriptor: ROOT_DESCRIPTOR,
     descMap: getDescriptorMap()
 }
-service /HelloWorld on new grpc:Listener(9090) {
+service "HelloWorld" on ep {
+    remote function hello(HelloWorldStringCaller caller, ContextString request) {
+        io:println("Invoked the hello RPC call.");
+        // Reads the request content
```
Comment (praneesha) [inline examples/grpc-unary-blocking/grpc_unary_blocking_service.bal:14]:
```suggestion
        // Reads the request content.
```

File: examples/grpc-unary-blocking/grpc_unary_blocking_service.bal [hunk 43]
```diff
@@ -1,36 +1,58 @@
...
     descMap: getDescriptorMap()
 }
-service /HelloWorld on new grpc:Listener(9090) {
+service "HelloWorld" on ep {
+    remote function hello(HelloWorldStringCaller caller, ContextString request) {
+        io:println("Invoked the hello RPC call.");
+        // Reads the request content 
+        string message = "Hello " + request.content;
+        io:println(request);
+        // Reads the custom headers in the request request.headers.hasKey
```
Comment (praneesha) [inline examples/grpc-unary-blocking/grpc_unary_blocking_service.bal:17]:
```suggestion
        // Reads the custom headers in the request as `request.headers.hasKey`.
```

File: examples/grpc-unary-blocking/grpc_unary_blocking_service.bal [hunk 44]
```diff
@@ -1,36 +1,58 @@
...
+        string[] reqHeaders = request.headers.hasKey("client_header_key")?request.headers.get("client_header_key"):["none"];
+        io:println("Server received header value: " + reqHeaders[0]);
+        // Writes custom headers to response message.
```
Comment (praneesha) [inline examples/grpc-unary-blocking/grpc_unary_blocking_service.bal:20]:
```suggestion
        // Writes the custom headers to the response message.
```

File: examples/grpc-unary-blocking/grpc_unary_blocking_service.bal [hunk 45]
```diff
@@ -1,36 +1,58 @@
...
+        map<string[]> responseHeaders = {};
+        responseHeaders["server_header_key"] = ["Response Header value"];
+        // Set up the response message and send it
```
Comment (praneesha) [inline examples/grpc-unary-blocking/grpc_unary_blocking_service.bal:23]:
```suggestion
        // Set up the response message and send it.
```

File: examples/grpc-unary-blocking/grpc_unary_blocking_service.bal [hunk 46]
```diff
@@ -1,36 +1,58 @@
...
-        if (result is grpc:Error) {
-            log:printError("Error in sending completed notification to caller",
-                err = result);
-        }
+    isolated remote function complete() returns grpc:Error? {
+        return self.caller->complete();
     }
 }
+
+# Context record includes message payload and headers.
```
Comment (praneesha) [inline examples/grpc-unary-blocking/grpc_unary_blocking_service.bal:54]:
```suggestion
# The `context` record includes the message payload and headers.
```

File: examples/grpc-unary-blocking/grpc_unary_blocking_service_client.bal [hunk 47]
```diff
@@ -2,27 +2,65 @@
...
-    } else {
-        string result;
-        grpc:Headers resHeaders;
-        [result, resHeaders] = unionResp;
-        io:println("Client Got Response : " + result);
-        string headerValue = resHeaders.get("server_header_key") ?: "none";
-        io:println("Headers: " + headerValue);
+public function main (string... args) returns error? {
+    HelloWorldClient ep = check new("http://localhost:9090");
+    //Setting the custom headers
```
Comment (praneesha) [inline examples/grpc-unary-blocking/grpc_unary_blocking_service_client.bal:7]:
```suggestion
    //Setting the custom headers.
```

File: examples/grpc-unary-blocking/grpc_unary_blocking_service_client.bal [hunk 48]
```diff
@@ -2,27 +2,65 @@
...
+    map<string[]> requestHeaders = {};
+    requestHeaders["client_header_key"] = ["0987654321"];
+    ContextString requestMessage = {content: "WSO2", headers: requestHeaders};
+    // Executing unary call
```
Comment (praneesha) [inline examples/grpc-unary-blocking/grpc_unary_blocking_service_client.bal:11]:
```suggestion
    // Executing the unary call.
```

File: examples/grpc-unary-blocking/grpc_unary_blocking_service_client.bal [hunk 49]
```diff
@@ -2,27 +2,65 @@
...
+    ContextString result = check ep->helloContext(requestMessage);
+    // Print the content
```
Comment (praneesha) [inline examples/grpc-unary-blocking/grpc_unary_blocking_service_client.bal:13]:
```suggestion
    // Print the content.
```

File: examples/grpc-unary-blocking/grpc_unary_blocking_service_client.bal [hunk 50]
```diff
@@ -2,27 +2,65 @@
...
+    io:println(result.content);
+    // Print a header value
```
Comment (praneesha) [inline examples/grpc-unary-blocking/grpc_unary_blocking_service_client.bal:15]:
```suggestion
    // Print a header value.
```

File: examples/grpc-unary-blocking/grpc_unary_blocking_service_client.bal [hunk 51]
```diff
@@ -2,27 +2,65 @@
...
+}
+
+public client class HelloWorldClient {
+
+    *grpc:AbstractClientEndpoint;
+
+    private grpc:Client grpcClient;
+
+    public isolated function init(string url, grpc:ClientConfiguration? config = ()) returns grpc:Error? {
+        // Initialize client endpoint.
```
Comment (praneesha) [inline examples/grpc-unary-blocking/grpc_unary_blocking_service_client.bal:26]:
```suggestion
        // Initialize the client endpoint.
```

File: examples/grpc-unary-blocking/grpc_unary_blocking_service_client.bal [hunk 52]
```diff
@@ -2,27 +2,65 @@
...
+            message = req;
+        }
+        var payload = check self.grpcClient->executeSimpleRPC("HelloWorld/hello", message, headers);
+        [anydata, map<string[]>][result, respHeaders] = payload;
+        return {content: result.toString(), headers: respHeaders};
     }
+
 }
+
+# Context record includes message payload and headers.
```
Comment (praneesha) [inline examples/grpc-unary-blocking/grpc_unary_blocking_service_client.bal:62]:
```suggestion
# The `context` record includes the message payload and headers.
```

File: examples/grpc-client-streaming/grpc_client_streaming_service.bal (hunk 1 above)
Comment (BuddhiWathsala) [reply examples/grpc-client-streaming/grpc_client_streaming_service.bal]: Fixed in here https://github.com/ballerina-platform/ballerina-distribution/pull/1180/commits/05e21f2d92ca432dc94ef290c343eb509c05ef7e

**Review by praneesha — APPROVED** [review]

File: examples/grpc-client-streaming/grpc_client_streaming_service.bal [hunk 53]
```diff
@@ -1,33 +1,54 @@
-// This is the server implementation for the client streaming scenario.
+// This is the server implementation of the client streaming scenario.
```
Comment (ThisaruGuruge) [inline examples/grpc-client-streaming/grpc_client_streaming_service.bal:1]: Do we need this line? The BBE title expresses the same idea right?

Comment (BuddhiWathsala) [reply examples/grpc-client-streaming/grpc_client_streaming_service.bal:1]: It is to differentiate the proto definition, client implementation, and server implementation.

**Review by ThisaruGuruge — APPROVED** [review]

**Review by daneshk — APPROVED** [review]

**Pull request merged**